### PR TITLE
BRS bugs and feature improvements. 

### PR DESCRIPTION
**Pull request by ayushi1411**
1. When Clicking on backup GUID on list backups, it opens a json page (SFX) https://github.com/microsoft/service-fabric-explorer/issues/461
image
the issue fixed. On clicking the backup id GUID, the backup details will be displayed. 
image

2.  When Clicking on backup GUID on list backups, it opens up blank dialog box (SFX - Beta) https://github.com/microsoft/service-fabric-explorer/issues/461
image
The issue is fixed and now displays the backup details
image

3. Implemented a new feature to enable the customer to view the entities enabled by a backup policy via SFX itself instead of the customer making a restAPI call. The list supports pagination, search and export. https://github.com/microsoft/service-fabric-explorer/issues/463
image
image

4. customers will now have the option to restore the latest backup taken directly from SFX https://github.com/microsoft/service-fabric-explorer/issues/462
image

5. Customers can also restore the partition using any of the backups through SFX https://github.com/microsoft/service-fabric-explorer/issues/462
image

6. Added info for inputs while creating and updating backup policy to enable customers to know the format of the input instead of browsing through the documentation.  https://msazure.visualstudio.com/One/_workitems/edit/7640655
image
image
image

7. Customers were unable to trigger **backup partition** and **restore partition** through SFX. Necessary fixes done to resolve the issue
https://github.com/microsoft/service-fabric/issues/1092
https://msazure.visualstudio.com/One/_workitems/edit/8651094

8. Date and time filter were not present for backup list in SFX-BETA. The date and time filters have been added and it was made sure that the issues faced with the date and time filter in SFX is not faced in SFX-BETA.
https://github.com/microsoft/service-fabric-explorer/issues/460
image
image

### CODE REVIEW

**Review by mevora-msft — APPROVED** [review]

File: src/SfxWeb/src/app/modules/backup-restore/get-backup-enabled-entities/get-backup-enabled-entities.component.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  response = '';
+  backupEnabledEntitiesInfoListSettings: ListSettings;
+  constructor(public dialogRef: MatDialogRef<GetBackupEnabledEntitiesComponent>,
+              @Inject(MAT_DIALOG_DATA) public data: IsolatedAction, public dataService: DataService) {
+     }
+
+  ngOnInit(){
+    this.rawdata = this.data.data.backupEntity;
+    this.backupEnabledEntitiesInfoListSettings = new ListSettings(10, null, [
+      new ListColumnSetting('EntityKind', 'Entity Kind', {colspan: 10}),
```
Comment (jeffj6123) [inline src/SfxWeb/src/app/modules/backup-restore/get-backup-enabled-entities/get-backup-enabled-entities.component.ts:32]: whats the reason for the colspans?

File: src/SfxWeb/src/app/shared/component/dual-date-picker/dual-date-picker.component.ts [hunk 2]
```diff
@@ -49,13 +49,12 @@ export class DualDatePickerComponent implements OnInit, OnChanges {
   onDateSelection(date: NgbDate) {
     if (!this.fromDate && !this.toDate) {
       this.fromDate = date;
-    } else if (this.fromDate && !this.toDate && date.after(this.fromDate)) {
+    } else if (this.fromDate && !this.toDate && !date.before(this.fromDate)) {
```
Comment (jeffj6123) [inline src/SfxWeb/src/app/shared/component/dual-date-picker/dual-date-picker.component.ts:52]: why this change?

File: src/SfxWeb/src/app/views/partition/backups/backups.component.ts [hunk 3]
```diff
@@ -23,12 +25,21 @@ export class BackupsComponent extends PartitionBaseControllerDirective {
...
+  minDate: Date;
+  maxDate: Date;
+  startTime: any;
+  endTime: any;
+  private debouncerHandlerSubscription: Subscription;
   constructor(protected data: DataService, injector: Injector, private settings: SettingsService, public telemetry: TelemetryService) {
     super(data, injector);
   }
+  backupList: any;
+  dateRefresh: boolean;
```
Comment (jeffj6123) [inline src/SfxWeb/src/app/views/partition/backups/backups.component.ts:40]: whats the reason for only refreshing once?

File: src/SfxWeb/src/app/views/partition/backups/backups.component.ts [hunk 4]
```diff
@@ -41,14 +52,88 @@ export class BackupsComponent extends PartitionBaseControllerDirective {
...
+  endTimeChange(){
+    const time = this.endTime.split(':');
+    this.endDate.setUTCHours(parseInt(time[0], 10), parseInt(time[1], 10), parseInt(time[2], 10), 0);
+    this.setNewPartitionBackupList(this.startDate, this.endDate);
+  }
+
+  setNewPartitionBackupList(startDate: Date, endDate: Date)
+  {
+    this.backupList = this.partition.partitionBackupInfo.partitionBackupList.collection;
+    this.backupList = this.backupList.sort((left, right): number => {
```
Comment (jeffj6123) [inline src/SfxWeb/src/app/views/partition/backups/backups.component.ts:82]: why sort here instead of letting the list display component handle it?

File: src/SfxWeb/src/app/views/partition/backups/backups.component.ts [hunk 5]
```diff
@@ -41,14 +52,88 @@ export class BackupsComponent extends PartitionBaseControllerDirective {
...
+      if (left.CreationTimeUtc < right.CreationTimeUtc) {
+        return -1;
+      }
+      if (left.CreationTimeUtc > right.CreationTimeUtc) {
+        return 1;
+      }
+      return 0;
+    });
+    this.backupList = this.backupList.filter(function getData(info) {
```
Comment (jeffj6123) [inline src/SfxWeb/src/app/views/partition/backups/backups.component.ts:91]: use an arrow function for consistency here

File: src/SfxWeb/src/app/views/partition/backups/backups.component.ts [hunk 6]
```diff
@@ -41,14 +52,88 @@ export class BackupsComponent extends PartitionBaseControllerDirective {
...
     if (this.data.actionsEnabled()) {
       this.attemptSetActions();
     }
-
+    if (this.dateRefresh)
+    {
+      this.backupList = this.partition.partitionBackupInfo.partitionBackupList.collection;
+      if (this.backupList.length !== 0)
+      {
+        const templist = this.backupList.sort((left, right): number => {
```
Comment (jeffj6123) [inline src/SfxWeb/src/app/views/partition/backups/backups.component.ts:119]: why sort here instead of in list view component.

File: src/SfxWeb/src/app/views/partition/backups/backups.component.ts [hunk 7]
```diff
@@ -41,14 +52,88 @@ export class BackupsComponent extends PartitionBaseControllerDirective {
...
+          if (left.CreationTimeUtc < right.CreationTimeUtc) {
+            return -1;
+          }
+          if (left.CreationTimeUtc > right.CreationTimeUtc) {
+            return 1;
+          }
+          return 0;
+        });
+        this.startDate = new Date(templist[0].raw.CreationTimeUtc);
```
Comment (jeffj6123) [inline src/SfxWeb/src/app/views/partition/backups/backups.component.ts:128]: do all of the date ranges get reset on a page refresh?

File: src/SfxWeb/src/app/views/partition/backups/backups.component.ts [hunk 8]
```diff
@@ -41,14 +52,88 @@ export class BackupsComponent extends PartitionBaseControllerDirective {
       new ListColumnSetting('raw.LsnOfLastBackupRecord', 'Lsn of last Backup Record'),
       new ListColumnSetting('raw.CreationTimeUtc', 'Creation Time UTC'),
     ]);
-    console.log(this);
+    this.debouncerHandlerSubscription = this.debounceHandler
+    .pipe(debounceTime(400), distinctUntilChanged())
```
Comment (jeffj6123) [inline src/SfxWeb/src/app/views/partition/backups/backups.component.ts:56]: whats the benefit for using this? Is there a network call involved when filtering the list?

File: src/SfxWeb/src/app/views/partition/backups/backups.component.html [hunk 9]
```diff
@@ -27,7 +30,17 @@ <h4>Partition Restore Progress</h4>
...
     <div *ngIf="partition && partition.replicas.isInitialized" class="detail-pane essen-pane">
-        <app-detail-list [list]="partition.partitionBackupInfo.partitionBackupList.collection" [listSettings]="partitionBackupListSettings" ></app-detail-list>
+        <app-detail-list [list]="backupList" [listSettings]="partitionBackupListSettings" >
+            <app-dual-date-picker [minDate]="minDate" [maxDate]="maxDate"
+            [currentStartDate]="startDate" [currentEndDate]="endDate" (dateChanged)="setNewDates($event)">
+            <div startTime>
+                <input type="time" [(ngModel)]="startTime" placeholder="HH:mm:ss" class="input-flat clear-disabled" ng-model-options="{ timezone: 'UTC' }" (change)="startTimeChange()" />
+            </div>
+            <div toTime>
+                <input type="time" [(ngModel)]="endTime" placeholder="HH:mm:ss" class="input-flat clear-disabled" ng-model-options="{ timezone: 'UTC' }" (change)="endTimeChange()" />
```
Comment (jeffj6123) [inline src/SfxWeb/src/app/views/partition/backups/backups.component.html:40]: is it necessary to use another variable to track the h/m/s part of the time? would using startDate/EndDate for these pickers have any issues? Shouldnt they just update that part of the time?

File: src/SfxWeb/src/app/shared/component/dual-date-picker/dual-date-picker.component.ts (hunk 2 above)
Comment (ayushi1411) [reply src/SfxWeb/src/app/shared/component/dual-date-picker/dual-date-picker.component.ts:52]: before the change, the calender was not allowing to choose the same start and end dates, which should be possible to do. I should be able to filter choosing a particular date

File: src/SfxWeb/src/app/views/partition/backups/backups.component.ts (hunk 3 above)
Comment (ayushi1411) [reply src/SfxWeb/src/app/views/partition/backups/backups.component.ts:40]: thanks Jeffrey. I want the calendar to reset only when i refresh the page, instead of every time the list is refreshed. The refresh method seems to be called every seconds and hence was unable to filter out the list based on date and time

File: src/SfxWeb/src/app/shared/component/dual-date-picker/dual-date-picker.component.ts (hunk 2 above)
Comment (jeffj6123) [reply src/SfxWeb/src/app/shared/component/dual-date-picker/dual-date-picker.component.ts:52]: sounds good, was not aware this was an issue.

File: src/SfxWeb/src/app/views/partition/backups/backups.component.ts (hunk 3 above)
Comment (jeffj6123) [reply src/SfxWeb/src/app/views/partition/backups/backups.component.ts:40]: Do you want it to function similar to how event store does tab does?

Comment (ayushi1411) [reply src/SfxWeb/src/app/views/partition/backups/backups.component.ts:40]: yes. But the backup list is just a section of the backup tab. I wanted to set the date and time in "setup()" but the list data is not accessible there and I think from what i found is that ngOnInit() and setup() didn't work together

File: src/SfxWeb/src/app/views/partition/backups/backups.component.ts (hunk 6 above)
Comment (ayushi1411) [reply src/SfxWeb/src/app/views/partition/backups/backups.component.ts:119]: The reason for sorting the list here is to get the first and last timestamps in "CreationTimeUtc" . The "CreationTimeUtc" in the list is in string format so not sure if the max-min function would work.

File: src/SfxWeb/src/app/views/partition/backups/backups.component.ts (hunk 7 above)
Comment (ayushi1411) [reply src/SfxWeb/src/app/views/partition/backups/backups.component.ts:128]: date filter will get reset on manual page refresh. I ideally don't want the dates to refresh everytime the backup list is getting refreshed

File: src/SfxWeb/src/app/modules/backup-restore/get-backup-enabled-entities/get-backup-enabled-entities.component.ts (hunk 1 above)
Comment (ayushi1411) [reply src/SfxWeb/src/app/modules/backup-restore/get-backup-enabled-entities/get-backup-enabled-entities.component.ts]: Thanks Jeffrey, forgot to remove the colspans. Made the necessary changes

File: src/SfxWeb/src/app/views/partition/backups/backups.component.ts (hunk 5 above)
Comment (ayushi1411) [reply src/SfxWeb/src/app/views/partition/backups/backups.component.ts]: Thanks Jeffrey, Made the necessary changes

File: src/SfxWeb/src/app/views/partition/backups/backups.component.ts (hunk 4 above)
Comment (ayushi1411) [reply src/SfxWeb/src/app/views/partition/backups/backups.component.ts]: Sorting the list here is unnecessary, made the necessary changes

File: src/SfxWeb/src/app/views/partition/backups/backups.component.ts (hunk 8 above)
Comment (ayushi1411) [reply src/SfxWeb/src/app/views/partition/backups/backups.component.ts]: Thanks Jeffrey, I had made these changes by referring to event store component. I have made the necessary changes in the code.

File: src/SfxWeb/src/app/views/partition/backups/backups.component.html (hunk 9 above)
Comment (ayushi1411) [reply src/SfxWeb/src/app/views/partition/backups/backups.component.html:40]: input value goes to startTime/endTime as a string. if I use startDate/endDate in ngModel, it overrides the entire date with the time string.

**Review by jeffj6123 — APPROVED** [review]
